tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -0,0 +1,321 @@
+  Value memref = result_memref;
+  if (shape_constraint_analysis) {
+    Value leader_memref =
+        shape_constraint_analysis->GetLeaderValueWithSameShape(result_memref);
+    if (leader_memref != nullptr) {
+      memref = leader_memref;
+    }
+  }
+  auto output_multidim_index =
+      calcMultiDimIndex(b, loc, output_linear_index, memref);
+  auto operand_data =
+      elementalLower<LHLO_OpTy>(&b, loc, op, output_multidim_index,
+                                /*check_cache*/ true);
+  createOffsetStore(b, loc, operand_data, result_memref, output_linear_index);
+  return true;
+}
+
+template <typename First>
+bool elemwiseLowerHelperOr(
+    OpBuilder& b, Location loc, Operation* op, Value output_linear_index,
+    const ShapeConstraintAnalysis* shape_constraint_analysis) {
+  return elemwiseLowerHelper<First>(b, loc, op, output_linear_index,
+                                    shape_constraint_analysis);
+}
+
+template <typename First, typename Second, typename... Rest>
+bool elemwiseLowerHelperOr(
+    OpBuilder& b, Location loc, Operation* op, Value output_linear_index,
+    const ShapeConstraintAnalysis* shape_constraint_analysis) {
+  return elemwiseLowerHelperOr<First>(b, loc, op, output_linear_index,
+                                      shape_constraint_analysis) ||
+         elemwiseLowerHelperOr<Second, Rest...>(b, loc, op, output_linear_index,
+                                                shape_constraint_analysis);
+}
+
+LogicalResult lowerHelper(
+    OpBuilder& b, Location loc, Operation* op, Value output_linear_index,
+    const ShapeConstraintAnalysis* shape_constraint_analysis) {
+  if (elemwiseLowerHelperOr<
+#define GET_ELEMENTWISE_OP_LIST
+#include "tensorflow/compiler/mlir/hlo/lib/utils/elementwise_op_list.cc"
+          >(b, loc, op, output_linear_index, shape_constraint_analysis) ||
+      // TODO: Upstream is on the way for more Ops
+      miscLowerHelper<RealDynamicSliceOp>(b, loc, op, output_linear_index,
+                                          shape_constraint_analysis) ||
+      miscLowerHelper<DynamicBroadcastInDimOp>(b, loc, op, output_linear_index,
+                                               shape_constraint_analysis) ||
+      miscLowerHelper<BroadcastInDimOp>(b, loc, op, output_linear_index,
+                                        shape_constraint_analysis)) {
+    return success();
+  } else {
+    assert(false && "unsupported lhlo_op");
+    llvm::dbgs() << "unsupported lhlo_op: " << op->getName();
+    return failure();
+  }
+}
+
+// we don't do inbound check for kLoop Schedule
+// LoopSplit pass will do this.
+//
+/* %num_elements = ElementsIn(root_shape)
+ * loop.for %idx = 0 to %num_elements step 1 {
+ *   %multidim_indices_0..n = getMultidimIndices(%idx);
+ *   %operand_0 = load %operand0[]
+ *   %operand_1 = load %operand1[]
+ *   emit calculation..
+ * }
+ */
+LogicalResult lowerWithScheduleLoop(
+    OpBuilder& b, ArrayRef<Operation*> root_ops, Operation* dominant_op,
```

Comment:
Why couple "creating the loop" with "lowering root ops"? Can't we create an empty loop and then let InputInlineFusion bring all ops (including root ops) into the loop?

Ideally we should be able to unify `InputInlineFusionPattern::inlineFuseLhloOp` and `lowerHelper`, right?

Reply:
For kLoop fusion, i think it's possible. The reason to seperate the two phase is for covering kInput fusions (not submitted yet in this PR), in which case the root node codegen will be quite different.